src/main/scala/viper/gobra/translator/encodings/interfaces/TypeComponentImpl.scala

```diff
@@ -267,12 +269,19 @@ class TypeComponentImpl extends TypeComponent {
         formalArgs = varsDecl,
         typ = domainType
       )(domainName = domainName)
-
       val funcApp = vpr.DomainFuncApp(func = func, vars, Map.empty)()
 
+      val tagValueFunc = vpr.DomainFunc(
+        name = s"${functionName(name)}_tag",
+        formalArgs = Seq.empty,
+        typ = vpr.Int,
+        unique = true
+      )(domainName = domainName)
+      val tagValueFuncApp = vpr.DomainFuncApp(func = tagValueFunc, Seq.empty, Map.empty)()
+
       val tagAxiom = {
         val tagApp = vpr.DomainFuncApp(func = tagFunc, Seq(funcApp), Map.empty)()
-        val eq = vpr.EqCmp(tagApp, vpr.IntLit(tag)())()
+        val eq = vpr.EqCmp(tagApp, tagValueFuncApp)()
```

Comment:
This feels a bit weird. Could you not just collect all functions necessary and then generate them at the finalizing step.

Reply:
Yes, I am worried about performance.

Reply:
if there is no performance drop, then you can merge.